crates/tx5/src/lib.rs

```diff
@@ -36,261 +36,177 @@ pub mod deps {
-                let mut item = match self.0.pop_front() {
-                    Some(item) => item,
-                    None => panic!("UnexpectedEof"),
-                };
-
-                let rem = item.remaining();
-                if rem == cnt {
-                    out.push(item);
-                    return out;
-                } else if rem < cnt {
-                    out.push(item);
-                    cnt -= rem;
-                } else if rem > cnt {
-                    out.push(item.split_to(cnt));
-                    self.0.push_front(item);
-                    return out;
-                }
-            }
-        }
-    */
 }
 
-/*
-impl From<std::collections::VecDeque<bytes::Bytes>> for BytesList {
-    #[inline(always)]
-    fn from(v: std::collections::VecDeque<bytes::Bytes>) -> Self {
-        Self(v)
-    }
+/// Make a shared (clonable) future abortable and set up an automatic
+/// abort at a time in the future specified by a duration.
+struct AbortableTimedSharedFuture<T: Clone> {
+    f: futures::future::Shared<
+        futures::future::BoxFuture<'static, std::result::Result<T, Error>>,
+    >,
+    a: std::sync::Arc<
+        std::sync::Mutex<Option<tokio::sync::oneshot::Sender<Error>>>,
+    >,
+    t: DoDrop,
 }
 
-impl From<bytes::Bytes> for BytesList {
-    #[inline(always)]
-    fn from(b: bytes::Bytes) -> Self {
-        let mut out = std::collections::VecDeque::with_capacity(8);
-        out.push_back(b);
-        out.into()
+impl<T: Clone> Clone for AbortableTimedSharedFuture<T> {
+    fn clone(&self) -> Self {
+        Self {
+            f: self.f.clone(),
+            a: self.a.clone(),
+            t: self.t.clone(),
+        }
     }
 }
 
-impl From<Vec<u8>> for BytesList {
-    #[inline(always)]
-    fn from(v: Vec<u8>) -> Self {
-        bytes::Bytes::from(v).into()
+impl<T: Clone> AbortableTimedSharedFuture<T> {
+    /// Construct a new AbortableTimedSharedFuture that will timeout
+    /// after the given duration.
+    pub fn new<F>(
+        timeout: std::time::Duration,
+        timeout_err: Error,
+        f: F,
+    ) -> Self
+    where
+        F: std::future::Future<Output = std::result::Result<T, Error>>
+            + 'static
+            + Send,
+    {
+        let (a, ar) = tokio::sync::oneshot::channel();
+        let a = std::sync::Arc::new(std::sync::Mutex::new(Some(a)));
+        let a2 = a.clone();
+        let t = DoDrop(std::sync::Arc::new(tokio::task::spawn(async move {
+            tokio::time::sleep(timeout).await;
+            do_abort(&a2, timeout_err);
+        })));
+        Self {
+            f: futures::future::FutureExt::shared(
+                futures::future::FutureExt::boxed(async move {
+                    tokio::select! {
+                        r = async {
+                            Err(ar.await.map_err(|_| Error::id("AbortHandleDropped"))?)
+                        } => r,
+                        r = f => r,
+                    }
+                }),
```

Comment:
I wonder if plain ol' tokio::time::timeout would have worked too? Just a task that awaits the timeout of the input future and maps the timeout error to the custom error? This is fine too but seems like a reinvention of timeout?

Reply:
done: https://github.com/holochain/tx5/pull/77/commits/4a24d8b6c0eff1218834dd8d0dd4971ed051602f

Yeah, I added the timeout bit after just having sorted out the aborting logic, so since I was holding a hammer, I hammered it... tokio::time::timeout is much better, thanks for pointing that out.

Also, I'm pretty sure the drop logic was just plain wrong. If one holochain task wanted to send some data but then got aborted, the drop logic would have aborted any future tasks that had tried to send data to the same endpoint even though they might have still wanted to send their data... so I got rid of that.

Reply:
Looks good!